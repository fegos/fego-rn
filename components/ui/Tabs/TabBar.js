import React from 'react';
import PropTypes from 'prop-types';
import {
  Text,
  View,
  ViewPropTypes,
  Animated,
  ScrollView,
  StyleSheet,
  TouchableWithoutFeedback,
} from 'react-native';
import merge from 'lodash/merge';
import { UIComponent } from 'common';

const tabStyle = {
  underline: {
    position: 'absolute',
    height: 4,
    bottom: 0,
  },
};


export default class TabBar extends UIComponent {
  static defaultProps = {
    activeIdx: null,
    duration: 350,
    showUnderline: true,
    style: null,
    tabStyle: null,
    textStyle: null,
    activeTextStyle: null,
    activeUnderlineStyle: null,
    onTabClick: () => { },
    onScrollEnd: () => { },
    onAnimationEnd: () => { },
  }

  static propTypes = {
    // tab 标签
    items: PropTypes.arrayOf(PropTypes.any).isRequired,
    // 当前激活的 tab 在 tabs 数组中的下标
    activeIdx: PropTypes.number,
    // tabBar宽度
    barWidth: PropTypes.number,
    // TabBarItem宽度
    itemWidth: PropTypes.number,
    // 动画持续时间
    duration: PropTypes.number,
    // 是否显示bottomline
    showUnderline: PropTypes.bool,
    // TabBar容器样式
    style: ViewPropTypes.style,
    // TabBarItem样式
    itemStyle: ViewPropTypes.style,
    // TabBarItem高亮样式
    activeItemStyle: ViewPropTypes.style,
    // title文字样式
    textStyle: Text.propTypes.style,
    // 文字高亮样式
    activeTextStyle: Text.propTypes.style,
    // 下划线高亮样式
    activeUnderlineStyle: ViewPropTypes.style,
    // 点击TabBarItem的回调
    onTabClick: PropTypes.func,
    // TabBar可滚动时，滚动后的回调，参数为滚动后的x方向上的偏移
    onScrollEnd: PropTypes.func,
    // TabBar选中变更动画结束回调
    onAnimationEnd: PropTypes.func,
  }

  static autoStyleSheet = false;

  constructor(props) {
    super(props);

    this.state = {
      tabBarWidth: props.barWidth,
      itemWidth: this._getItemWidth(props, props.barWidth),
      left: new Animated.Value(props.activeIdx * this._getItemWidth(props, props.barWidth)),
    };

    this._contentOffsetX = 0;
  }

  componentDidMount() {
    setTimeout(() => {
      this._animatedToIndex(this.props.activeIdx);
    }, 0);
  }

  componentWillReceiveProps(nextProps) {
    super.componentWillReceiveProps(nextProps);
    const index = nextProps.activeIdx;
    const { itemWidth, tabBarWidth } = this.state;
    const newItemWidth = this._getItemWidth(nextProps, tabBarWidth);
    if (itemWidth !== newItemWidth) {
      this.setState({
        itemWidth: newItemWidth,
      }, () => {
        this._animatedToIndex(index);
      });
    } else if (index !== this.props.activeIdx) {
      this._animatedToIndex(index, nextProps);
    }
  }

  /**
   * 更新item宽度
   *
   * @param {any} props
   * @returns
   * @memberof TabBar
   */
  _getItemWidth(props, tabBarWidth) {
    const { itemWidth, items } = props;
    if (itemWidth) {
      return itemWidth;
    } else {
      return items.length ? tabBarWidth / items.length : 0;
    }
  }

  /**
   * 点击item
   *
   * @memberof TabBar
   */
  _onItemPress = (item) => {
    const { onTabClick } = this.props;
    if (onTabClick) {
      onTabClick(item._key, item.title, item.index);
    }
  }

  /**
   * 停止拖拽
   *
   * @memberof TabBar
   */
  _onScrollEndDrag = (e) => {
    const { onScrollEnd } = this.props;
    if (onScrollEnd) {
      onScrollEnd(e.nativeEvent.contentOffset.x);
    }
  }


  /**
   * 动画滚到目标位置
   *
   * @memberof TabBar
   */
  _animatedToIndex = (index, nextProps = this.props) => {
    if (this._underlineAnimation) {
      this._underlineAnimation.stop();
      this._underlineAnimation = null;
    }
    const { duration, onAnimationEnd, showUnderline } = nextProps;
    if (showUnderline) {
      const { itemWidth } = this.state;
      this._underlineAnimation = Animated.timing(
        this.state.left,
        {
          toValue: index * itemWidth,
          duration,
        },
      ).start(() => {
        this._underlineAnimation = null;
        if (onAnimationEnd) {
          onAnimationEnd(index);
        }
      });
    }
    this._scrollTabBar(index, true);
  }

  _scrollTabBar(index, animated) {
    const { itemWidth, tabBarWidth } = this.state;
    const start = itemWidth * index;
    const end = itemWidth * (index + 1);

    if (start - this._contentOffsetX < 0) {
      this._scrollView.scrollTo({ x: start, y: 0, animated });
      this._contentOffsetX = start;
    }
    if (end - this._contentOffsetX > tabBarWidth) {
      const offsetx = end - tabBarWidth;
      this._scrollView.scrollTo({ x: offsetx, y: 0, animated });
      this._contentOffsetX = offsetx;
    }
  }


  /**
   * 处理布局
   *
   * @memberof TabBar
   */
  _handleLayout = (e) => {
    const { width } = e.nativeEvent;
    const { tabBarWidth, activeIdx } = this.state;
    if (width !== tabBarWidth) {
      this.setState({
        itemWidth: this._getItemWidth(this.props, width),
        tabBarWidth: width,
      }, () => {
        this._animatedToIndex(activeIdx);
      });
    }
  }

  /**
   * 渲染TabItem
   *
   * @param {any} item
   * @param {any} activeIdx
   * @returns
   * @memberof DefaultTabBar
   */
  _renderTabItem(item, activeIdx) {
    const {
      itemStyle, activeItemStyle, textStyle, activeTextStyle,
    } = this.props;
    const isActive = activeIdx === item.index;

    const actualItemStyle = isActive ? [itemStyle, activeItemStyle] : itemStyle;
    const itemTextStyle = isActive ? [textStyle, activeTextStyle] : textStyle;

    return (
      <TouchableWithoutFeedback
        style={{ flex: 1 }}
        key={item.index}
        onPress={() => { this._onItemPress(item); }}
      >
        <View style={[{ height: this.style.container.height }, actualItemStyle]}>
          <Text style={itemTextStyle}>{item.title}</Text>
        </View>
      </TouchableWithoutFeedback >
    );
  }

  render() {
    const {
      items, activeIdx, activeUnderlineStyle, barWidth, showUnderline,
    } = this.props;
    const {
      itemWidth,
    } = this.state;

    const finalLineStyle = merge({}, {
      width: itemWidth,
      left: this.state.left,
    }, tabStyle.underline, ...activeUnderlineStyle);

    let { width } = finalLineStyle;
    if (finalLineStyle.marginHorizontal) {
      width -= 2 * finalLineStyle.marginHorizontal;
    } else if (finalLineStyle.marginLeft) {
      width -= finalLineStyle.marginLeft;
    } else if (finalLineStyle.marginRight) {
      width -= finalLineStyle.marginRight;
    }

    finalLineStyle.width = width;

    const contentView = (
      <View style={{ flex: 1, flexDirection: 'row', alignItems: 'center' }}>
        {items.map(item => this._renderTabItem(item, activeIdx))}
        {showUnderline ? <Animated.View style={finalLineStyle}
        /> : null}
      </View>
    );
    return (
      <View
        style={[{ width: barWidth }, this.style.container]}
        onLayout={this.handleLayout}
      >
        <ScrollView
          horizontal
          ref={(s) => { this._scrollView = s; }}
          automaticallyAdjustContentInsets={false}
          showsHorizontalScrollIndicator={false}
          showsVerticalScrollIndicator={false}
          bounces={false}
          onScrollEndDrag={this._onScrollEndDrag}
          contentContainerStyle={
            [
              {
                position: 'relative',
              },
              {
                width: itemWidth * items.length,
                height: this.style.container.height,
              },
            ]}
        >
          {contentView}
        </ScrollView>
      </View >
    );
  }
}
