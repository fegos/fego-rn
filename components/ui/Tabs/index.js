/**
 * 标签页组件
 * @author esky 翟宇佳
 */
import React from 'react';
import PropTypes from 'prop-types';
import {
  View,
  Dimensions,
  ScrollView,
  StyleSheet,
} from 'react-native';
import { UIComponent } from '../../common';

import TabBar from './TabBar';
import TabPane from './TabPane';
import SceneComponent from './SceneComponent';


const screenWidth = Dimensions.get('window').width;

export default class Tabs extends UIComponent {
  static defaultProps = {
    showTabBar: true,
    tabBarPosition: 'top',
    onChange: () => { },
    onTabClick: () => { },
    onAnimationEnd: () => { },
    animated: true,
    swipeable: true,
    usePreload: true,
    singleSidePreloadCount: 1,
  }

  static propTypes = {
    // 受控属性，当前激活 tab 面板的 key，需配合 onChange 维护 activeKey 值
    activeKey: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.number,
    ]),
    // 非受控属性，默认激活的 tab 面板的 key
    defaultActiveKey: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.number,
    ]),
    // 是否显示切换动画
    animated: PropTypes.bool,
    // 是否显示tabBar
    showTabBar: PropTypes.bool,
    // 是否显示底部选中线
    showUnderline: PropTypes.bool,
    // tab 位置 top/bottom
    tabBarPosition: PropTypes.oneOf(['top', 'bottom']),
    // 是否可以滑动 tab 内容进行切换
    swipeable: PropTypes.bool,
    // 是否使用预加载
    usePreload: PropTypes.bool,
    // 单边预加载的数量
    singleSidePreloadCount: PropTypes.number,
    // 切换面板（手动点击标签或是滑动 tab 内容切换）的回调
    onChange: PropTypes.func,
    // 点击 tab 标签的回调
    onTabClick: PropTypes.func,
    // tab 切换动画结束后执行的回调
    onAnimationEnd: PropTypes.func,
  }

  // 简单样式
  static simpleStyleProps = {
    textColor: { name: 'text', attr: 'color' },
    activeTextColor: { name: 'activeText', attr: 'color' },
    underlineColor: { name: 'bar', attr: 'borderColor' },
    activeUnderlineColor: { name: 'activeUnderline', attr: 'backgroundColor' },
    activeBgColor: { name: 'activeItem', attr: 'backgroundColor' },
  }

  static autoStyleSheet = false;

  constructor(props) {
    super(props);

    const width = props.containerWidth || screenWidth;
    const { activeKey, defaultActiveKey, children } = this.props;
    // 默认激活的 tab
    let realActiveKey = activeKey === undefined ? defaultActiveKey : activeKey;
    this._updateTabBarItems(children);
    if (!realActiveKey && this._tabBarItems.length) {
      realActiveKey = this._tabBarItems[0]._key;
    }

    this.state = {
      containerWidth: width,
      curPageIdx: this._getIndex(realActiveKey),
    };

    this._loadedScenes = [realActiveKey];
  }


  /**
   * lifecylce
   *
   * @memberof Tabs
   */
  componentDidMount() {
    setTimeout(() => {
      this._scrollToRightPlace(this.state.curPageIdx, false);
    }, 0);
  }

  componentDidUpdate(prevProps) {
    super.componentDidUpdate(prevProps);

    const { activeKey, children } = this.props;
    this._updateTabBarItems(children);
    // 受控属性，由父元素告知 activeKey 来进行更新
    if (activeKey !== undefined && activeKey !== prevProps.activeKey) {
      const idx = this._getIndex(activeKey);
      this._setCurrentPage(idx);
    }
  }

  /**
   * 更新TabBarItems
   *
   * @memberof Tabs
   */
  _updateTabBarItems = (children) => {
    const tabBarItems = [];
    React.Children.forEach(children, (child, idx) => {
      tabBarItems.push({
        _key: child.key,
        index: idx,
        title: child.props.title,
      });
    });
    this._tabBarItems = tabBarItems;
  }

  /**
   * 根据index获取item
   *
   * @param {any} index
   * @returns
   * @memberof Tabs
   */
  _getItem(index) {
    if (this._tabBarItems) {
      return this._tabBarItems[index];
    }
    return null;
  }

  /**
   * 根据itemKey获取index
   *
   * @param {any} key
   * @returns
   * @memberof Tabs
   */
  _getIndex(key) {
    const tabBarItems = this._tabBarItems;
    let index = 0;
    for (const item of tabBarItems) {
      if (item._key === key) {
        break;
      } else {
        index++;
      }
    }
    return index;
  }

  /**
   * 设置当前页
   *
   * @param {any} idx
   * @memberof Tabs
   */
  _setCurrentPage(idx) {
    this.setState({
      curPageIdx: idx,
    }, () => {
      this._scrollToRightPlace(idx);
    });
  }

  /**
   * TabBar点击回调
   *
   * @memberof Tabs
   */
  _onTabClick = (key, title, index) => {
    const { activeKey, onChange, onTabClick } = this.props;
    const { curPageIdx } = this.state;

    // onTabClick 回调
    if (onTabClick) {
      onTabClick(key, title);
    }

    if (curPageIdx !== index) {
      // 使用的是非受控属性 defaultActiveKey，更新组件维护的值
      if (activeKey === undefined) {
        this._setCurrentPage(index);
      }
      // onChange 回调
      onChange(key, title);
    }
  }


  /**
   * TabBar滚动结束回调
   *
   * @memberof Tabs
   */
  _onTabBarScrollEnd = (offsetX) => {
    this._tabBarOffsetX = offsetX;
  }

  _scrollToRightPlace = (nextPageIdx, animated = this.props.animated) => {
    const { containerWidth } = this.state;
    const offset = nextPageIdx * containerWidth;
    if (this._scrollView) {
      this._scrollView.scrollTo({ x: offset, y: 0, animated });
    }
  }

  _updateSceneKeys(nextPageIdx) {
    const newKeys = [];
    const showedSceneKeys = this._showedSceneKeys;

    this._tabs.forEach((t) => {
      if (showedSceneKeys.indexOf(t.index) >= 0 || this._shouldRenderSceneKey(t.index, nextPageIdx)) {
        newKeys.push(t.index);
      }
    });

    this._showedSceneKeys = newKeys;
  }

  /**
   * 预加载管理
   *
   * @param {any} key
   * @returns
   * @memberof Tabs
   */
  _shouldLoadPage(key) {
    const { singleSidePreloadCount, usePreload } = this.props;
    if (usePreload) {
      const { curPageIdx } = this.state;
      const idx = this._getIndex(key);
      let shouldLoad = (idx <= (curPageIdx + singleSidePreloadCount) &&
        idx > (curPageIdx - singleSidePreloadCount));
      if (!shouldLoad) {
        shouldLoad = this._loadedScenes.indexOf(key) >= 0;
      }
      return shouldLoad;
    } else {
      return true;
    }
  }

  _shouldUpdatePage(idx) {
    const { singleSidePreloadCount, usePreload } = this.props;
    if (usePreload) {
      const { curPageIdx } = this.state;
      return (idx <= (curPageIdx + singleSidePreloadCount) &&
        idx > (curPageIdx - singleSidePreloadCount));
    } else {
      return true;
    }
  }


  /**
   * ScrollView回调
   *
   * @memberof Tabs
   */
  _onMomentumScrollBeginAndEnd = (e) => {
    const offsetX = e.nativeEvent.contentOffset.x;
    const { containerWidth, curPageIdx } = this.state;
    const { activeKey, onChange } = this.props;
    const index = Math.round(offsetX / containerWidth);

    if (curPageIdx !== index) {
      // 使用的是非受控属性 defaultActiveKey，更新组件维护的值
      if (activeKey === undefined) {
        this._setCurrentPage(index);
      }

      const { key, title } = this._getItem(index);
      onChange(key, title);
    }
  }

  /**
   * TabBar切换动画结束回调
   *
   * @memberof Tabs
   */
  _onAnimationEnd = (activeIdx) => {
    const { _key, title } = this._getItem(activeIdx);
    this.props.onAnimationEnd(_key, title);
  }


  /**
   * 渲染区
   *
   * @memberof Tabs
   */
  _handleLayout = (e) => {
    const { width } = e.nativeEvent.layout;
    const { curPageIdx, containerWidth } = this.state;

    if (Math.round(width) !== Math.round(containerWidth)) {
      this.setState({ containerWidth: width });
      setTimeout(() => {
        this._scrollToRightPlace(curPageIdx, false);
      }, 0);
    }
  }

  _renderTabBar = () => {
    const { showTabBar, tabBarPosition, showUnderline } = this.props;
    if (!showTabBar) return null;

    const { style } = this;
    const { curPageIdx, containerWidth } = this.state;

    const topPostion = tabBarPosition === 'top';
    const barBaseStyle = topPostion ? style.barTop : style.barBottom;
    const linePosition = topPostion ? { bottom: -1 } : { top: -1 };
    const activeUnderlineStyle = [style.activeUnderline, linePosition];

    const tabBarProps = {
      items: this._tabBarItems,
      activeIdx: curPageIdx,
      showUnderline,
      barWidth: containerWidth,
      itemWidth: style.tabBarItem.width,
    };

    return (
      <TabBar
        ref={(t) => { this._tabBar = t; }}
        style={[barBaseStyle, style.tabBar]}
        itemStyle={style.tabBarItem}
        activeItemStyle={style.activeItem}
        textStyle={style.text}
        activeTextStyle={style.activeText}
        activeUnderlineStyle={activeUnderlineStyle}
        {...tabBarProps}
        onTabClick={(key, title, index) => this._onTabClick(key, title, index)}
        onScrollEnd={offset => this._onTabBarScrollEnd(offset)}
        onAnimationEnd={activeIdx => this._onAnimationEnd(activeIdx)}
      />
    );
  }

  _renderScenes() {
    const { children } = this.props;
    const { containerWidth } = this.state;

    return children.map((child, idx) => {
      const key = `scene${idx}`;
      let realChild = child;
      if (!this._shouldLoadPage) {
        realChild = <View key={key} />;
      }

      return (
        <SceneComponent
          key={key}
          style={{ width: containerWidth }}
          shouldUpdated={this._shouldUpdatePage(idx)}
        >
          {realChild}
        </SceneComponent>
      );
    });
  }

  _renderContent() {
    const { swipeable } = this.props;
    return (
      <ScrollView
        ref={(s) => { this._scrollView = s; }}
        horizontal
        pagingEnabled
        directionalLockEnabled
        keyboardShouldPersistTaps="handled"
        keyboardDismissMode="on-drag"
        alwaysBounceVertical={false}
        automaticallyAdjustContentInsets={false}
        showsHorizontalScrollIndicator={false}
        scrollEventThrottle={16}
        scrollEnabled={swipeable}
        scrollsToTop={false}
        onMomentumScrollBegin={this._onMomentumScrollBeginAndEnd}
        onMomentumScrollEnd={this._onMomentumScrollBeginAndEnd}
      >
        {this._renderScenes()}
      </ScrollView>
    );
  }

  render() {
    const { tabBarPosition } = this.props;
    return (
      <View onLayout={this._handleLayout} style={[{ flex: 1 }, this.style.container]}>
        {tabBarPosition === 'top' && this._renderTabBar()}
        {this._renderContent()}
        {tabBarPosition === 'bottom' && this._renderTabBar()}
      </View>
    );
  }
}

Tabs.TabPane = TabPane;

Tabs.baseStyle = {
  // tab bar
  tabBar: {
    height: 42,
    flexDirection: 'row',
    borderColor: '#ddd',
  },
  barTop: {
    borderTopWidth: 0,
    borderBottomWidth: StyleSheet.hairlineWidth,
  },
  barBottom: {
    borderTopWidth: StyleSheet.hairlineWidth,
    borderBottomWidth: 0,
  },
  // 单个TabBarItem
  tabBarItem: {
    flex: 1,
    paddingBottom: 0,
    alignItems: 'center',
    justifyContent: 'center',
  },
  // TabBarItem内标题
  text: {
    fontSize: 15,
    color: '#333',
  },
  // 激活态TabBarItem内标题
  activeText: {
    color: '#108ee9',
  },
  // 激活态线条
  activeUnderline: {
    height: 2,
    backgroundColor: '#108ee9',
  },
  // 激活态TabBarItem背景色
  activeItem: {
    // backgroundColor: '#108ee9',
  },
};
