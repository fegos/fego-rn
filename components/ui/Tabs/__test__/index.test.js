import React from 'react';
import Renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import Tabs from '../index';
import TabBar from '../TabBar';

jest.useFakeTimers();
describe('Tabs Tests', () => {
  let content;

  beforeEach(() => {
    content = [];
    const tabs = ['tab1', 'tab2', 'tab3'];
    tabs.forEach((index, key) => {
      const tab = tabs[key];
      content.push(<Tabs.TabPane tab={tab} key={tab} />);
    });
  });

  it('renders correctly', () => {
    const tree = Renderer.create((
      <Tabs>
        {content}
      </Tabs>
    )).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('shallow function', () => {
    const tabs = (
      <Tabs defaultActiveKey="0">
        {content}
      </Tabs>
    );
    const wrapper = shallow(tabs);

    expect(wrapper.find(TabBar)).toHaveLength(1);
  });
});
