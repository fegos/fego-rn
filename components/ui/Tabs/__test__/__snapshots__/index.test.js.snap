// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs Tests renders correctly 1`] = `
<View
  onLayout={[Function]}
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "width": 750,
        },
        Object {
          "borderBottomWidth": 0.5,
          "borderColor": "#ddd",
          "borderTopWidth": 0,
          "flexDirection": "row",
          "height": 42,
        },
      ]
    }
  >
    <RCTScrollView
      automaticallyAdjustContentInsets={false}
      bounces={false}
      contentContainerStyle={
        Array [
          Object {
            "position": "relative",
          },
          Object {
            "height": 42,
            "width": 750,
          },
        ]
      }
      horizontal={true}
      onScrollEndDrag={[Function]}
      showsHorizontalScrollIndicator={false}
      showsVerticalScrollIndicator={false}
    >
      <View>
        <View
          style={
            Object {
              "alignItems": "center",
              "flex": 1,
              "flexDirection": "row",
            }
          }
        >
          <View
            accessible={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Array [
                Object {
                  "height": 42,
                },
                Array [
                  Object {
                    "alignItems": "center",
                    "flex": 1,
                    "justifyContent": "center",
                    "paddingBottom": 0,
                  },
                  Object {},
                ],
              ]
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#333",
                    "fontSize": 15,
                  },
                  Object {
                    "color": "#108ee9",
                  },
                ]
              }
            />
          </View>
          <View
            accessible={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Array [
                Object {
                  "height": 42,
                },
                Object {
                  "alignItems": "center",
                  "flex": 1,
                  "justifyContent": "center",
                  "paddingBottom": 0,
                },
              ]
            }
          >
            <Text
              style={
                Object {
                  "color": "#333",
                  "fontSize": 15,
                }
              }
            />
          </View>
          <View
            accessible={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Array [
                Object {
                  "height": 42,
                },
                Object {
                  "alignItems": "center",
                  "flex": 1,
                  "justifyContent": "center",
                  "paddingBottom": 0,
                },
              ]
            }
          >
            <Text
              style={
                Object {
                  "color": "#333",
                  "fontSize": 15,
                }
              }
            />
          </View>
          <View
            style={
              Object {
                "backgroundColor": "#108ee9",
                "bottom": -1,
                "height": 2,
                "left": 0,
                "position": "absolute",
                "width": 250,
              }
            }
          />
        </View>
      </View>
    </RCTScrollView>
  </View>
  <RCTScrollView
    alwaysBounceVertical={false}
    automaticallyAdjustContentInsets={false}
    directionalLockEnabled={true}
    horizontal={true}
    keyboardDismissMode="on-drag"
    keyboardShouldPersistTaps="handled"
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    pagingEnabled={true}
    scrollEnabled={true}
    scrollEventThrottle={16}
    scrollsToTop={false}
    showsHorizontalScrollIndicator={false}
  >
    <View>
      <View
        style={
          Object {
            "width": 750,
          }
        }
      >
        <View
          style={
            Object {
              "flex": 1,
            }
          }
          tab="tab1"
        />
      </View>
      <View
        style={
          Object {
            "width": 750,
          }
        }
      >
        <View
          style={
            Object {
              "flex": 1,
            }
          }
          tab="tab2"
        />
      </View>
      <View
        style={
          Object {
            "width": 750,
          }
        }
      >
        <View
          style={
            Object {
              "flex": 1,
            }
          }
          tab="tab3"
        />
      </View>
    </View>
  </RCTScrollView>
</View>
`;
